### ISSUE TITLE
Poke ActionScheduler_QueueRunner early

### ISSUE THREAD

**Issue by PaulSchiretz** [Prospress/action-scheduler#930]
Hi,

I'm using ActionScheduler to schedule a lot of actions, basically I'm processing attachments in WP. For my use case i schedule async_actions, most of them are triggered via hooks that are happening in the background like someone uploaded an attachment, but there are also actions that can be triggered via an ajax request by the user from the WP-Backend.

For now the system is running and does it's job very nicely!

**The only thing that is a bit annoying is:** if nothing is in the queue and `ActionScheduler_QueueRunner` is waiting for the next Cron_trigger, after an user interaction(were requests are enqueued) it can take up to 60seconds till the queue gets looked at to process any action. I would like to be able to somehow poke the `ActionScheduler_QueueRunner` to look if something needs to get done.

**Note:** I don't want to process those manual actions directly via a loopback request, i think it would be nice to keep the queue intact.

**Possible Solution:**
So my idea is to trigger the action_scheduler_run_queue hook via a loopback request to poke the ActionScheduler_QueueRunner via a call to `do_action( 'action_scheduler_run_queue', $context_identifier );` to look for new actions and get the queue running again.

```
**
 * Trigger an early loopback request.
 */
function request_early_runner() {

	$response = wp_remote_post( admin_url( 'admin-ajax.php' ), array(
		'method'      => 'POST',
		'timeout'     => 45,
		'redirection' => 5,
		'httpversion' => '1.0',
		'blocking'    => false,
		'headers'     => array(),
		'body'        => array(
			'action'     => 'request_early_runner',
			'eg_nonce' => wp_create_nonce( 'request_early_runner' ),
		),
		'sslverify' => apply_filters( 'https_local_ssl_verify', false ),
		'cookies'     => array(),
	) );
}



/**
 * Handle requests initiated by request_early_runner() and start a queue runner if the request is valid.
 */
function request_early_runner_callback() {

	if ( isset( $_POST['eg_nonce'] ) && wp_verify_nonce( $_POST['eg_nonce'], 'request_early_runner' ) ) {
		$context_identifier = array( 'WP Cron' );
		do_action( 'action_scheduler_run_queue', $context_identifier );
	}

	wp_die();
}
add_action( 'wp_ajax_nopriv_request_early_runner', 'request_early_runner_callback', 0 );
```

So the idea is that whenever i call `request_early_runner()`  the `action_scheduler_run_queue` hook is run and looks for actions, exactly as it does via a cron trigger.

**The questions i have now are:**

1. Are there any downsides to this approach?
2. Is it safe to do it like that and is there still only one(or the max concurrent_batches amount) `ActionScheduler_QueueRunner` running at a time?
3. should i use 'WP Cron' or 'Async Request' as context_identifier?
4. Is there a better way to handle that scenario?

I hope i was clear by what my use case is otherwise i can clarify it a bit more....

Looking forward to some hints!

All the best,
Paul

**Comment by vedanshujain** [comment]
Hey, @PaulSchiretz , I don't see any issues with your solution, it will be safe and fine as far as I can see. Although, there are in-fact better ways IMO to handle this scenario:

1. You can set a lower wait time for individual queue processing by using the filter `action_scheduler_lock_duration`. Default value is 60s, you can probably lower it to 5s or something that you are happy with.
2. If you want to trigger whenever by sending in a request, we already have a ajax-action called `async_request_queue_runner` which if you send a request to (basically, POST to `/admin-ajax.php?action=async_request_queue_runner` IIRC) and it will trigger an AS queue, so need to add a new endpoint.


I am closing the issue with this, but don't hesitate to reply back and we can re-open if there are more concerns.

**Comment by PaulSchiretz** [comment]
Thanks, sounds good! I'll go with your sugestions and very likely be fine for now!
